### ISSUE TITLE
Support relative include path in .clangd file

### ISSUE THREAD

**Issue by cloudhan** [clangd/clangd#1038]
It seems there is no mechanism to support relative path expansion. For example

```yaml
CompileFlags:
  Add:
  - -Iinclude/path/relative/to/.clangd
 ```

`.clangd` file can generally check into the repo, then the repo path will not be deterministic, adding a mechanism to dynamically determine the path will be greatly helpful.

**Comment by HighCommander4** [comment]
Doesn't the **compiler** perform relative path expansion at the time it processes the `-I` argument?

**Comment by cloudhan** [comment]
Then relative to what? When you open a subdirectory of a project that might have some .clangd files nested deep, and red squiggle occurs, I bless you have time to figure it out.

So when a relative path occurs in the configuration, where does it relative to should be made clear in the document. If it is relative to the current working directory where the clangd server starts up, then there should be a way to specifiy where you exactly want it to relative to, regardless of the execution environment of the clangd server instance.

**Comment by HighCommander4** [comment]
I expect it would be interpreted relative to the directory in which the compile command is invoked, i.e. the `"directory"` field of the entries in `compile_commands.json`.

**Comment by sam-mccall** [comment]
The compile command is basically a pair {working directory, argv}. When argv are parsed by the driver, relative paths are resolved against the working directory.

First: what are you really trying to do?
If you have no compilation database, you should use `compile_flags.txt` instead of `.clangd`, this will set the working directory (as it sets rather than modifies the flags).
Are you trying to patch the compile commands from an existing CDB, to add extra includes?

The `CompileCommands` section in `.clangd` file modifies the argv textually, we can't reasonably say "relative paths are resolved against `.clangd` where they're specified, as the config parser doesn't even know which args specify paths (they're just strings).

Things we could possibly do:
 - provide a semantic config option like `AddIncludeDirectory: foo` that would be equivalent to `CompileFlags: Add: -I/absolute/path/foo`.
 - provide a config option to override the working directory for the compile command (though this would mess up relative paths specified in *other* places)
 - support some textual expansion like in your example

**Comment by cloudhan** [comment]
Sometimes I don't have CBD, because it requires extra work or the build system does not support generating it. I do it for patching a subdirectory (for example, opencl code). I find a `.clangd` file can be the last resort and can be very useful in this case. But not that useful if I cannot check it into the VCS. So to checkin, the relative path should be properly handled first.

So I think textual expansion just fit in this case. And it can be simplified as
```yaml
CompileFlags:
  Add:
  - -I${__FILE__}/../include/path/relative/to/.clangd
```
I think this change will be minimal. After loading, expand every string value and then form a static config.

**Comment by cloudhan** [comment]
`compile_flags.txt` doesn't fit the case because I use `If` block to pattern match files.

Anyway, If someone can finialize a design, I can implement it myself and contribute a PR :)

**Comment by dabeibao** [comment]
Really need this if compilation database is not available.
Otherwise looks like need to adjust .clang file every time I move code to another location.

**Comment by fneddy** [comment]
+1

**Comment by marioc8** [comment]
+1

**Comment by adenisonafifi** [comment]
Would appreciate this functionality. For now I have added a bash script to generate the `.clangd` file for each user so the include paths are correct for their repository checkout path. Would be great to be able to commit the `.clangd` file to the repo and avoid requiring a setup script.

**Comment by freqmod** [comment]
I made a review for a solution to this https://reviews.llvm.org/D142981

**Comment by miguno** [comment]
+1 to what @adenisonafifi said in the comment at https://github.com/clangd/clangd/issues/1038#issuecomment-1326557891

**Comment by bullno1** [comment]
> I expect it would be interpreted relative to the directory in which the compile command is invoked, i.e. the `"directory"` field of the entries in `compile_commands.json`.

Thanks.
At least that's solved it for me.

The problem is that the path is unintuitive.
Without a compilation database, it's relative to the .clangd file.
With it, it's relative to the `compile_commands.json` file.
And if you use CMake, it's usually a subdirectory.

**Comment by absop** [comment]
> I made a review for a solution to this https://reviews.llvm.org/D142981

Does this method also work for `compile_commands.json`? I have a `compile_commands.json` generated in Linux and I want it to work in Windows. I don't want to have to modify the common path of for the `"directory"` entries every time.

**Comment by HighCommander4** [comment]
> Does this method also work for `compile_commands.json`? I have a `compile_commands.json` generated in Linux and I want it to work in Windows. I don't want to have to modify the common path of for the `"directory"` entries every time.

No, `compile_commands.json` files are meant to be generated at build time on the machine on which they're consumed. They're not portable in any meaningful sense.

**Comment by curVV** [comment]
Is this possible yet?

**Comment by appthumb** [comment]
+1 on having relative path supported in .clangd directly, e.g., relative to the location of .clangd.

**Comment by ryukinix** [comment]
It would be interesting to have this and make the config more portable.

**Comment by HighCommander4** [comment]
I recently came across a workaround for this which may be of interest to folks following this issue.

The workaround applies in situations where you're not using `compile_commands.json`: merely placing an empty `compile_flags.txt` file in the project's root directory will have the effect of all compile commands (including flags coming from `Add:` keys in `.clangd` files) being interpreted relative to that directory.
